worker/updateChorals.js

```diff
@@ -2,42 +2,138 @@ var client = require('redis').createClient(process.env.REDIS_STORE_URI)
+        if(err) return console.log(err.message)
+
+        // json comes in as a string, parse it first
+        param[child.name] = {}
+        param[child.name].data = JSON.parse(data.device_data)
+        param[child.name].device_timestamp = data.device_timestamp | 0
+        // this is just so I dont have to import a promise library
+        if(Object.keys(param).length == choral.children.length){
+          // Amazon lambda functions would make this a viable solution to
+          // our problem because they do not have access to our environment
+          // if this ends up being super slow we should do that
+          runComputation(param, choral)
+        }
       })
     })
+}
+
+function startNewChorals(chorals){
+  chorals.forEach((choral) => {
+    // dont rerun already running chorals
+    if(TIMEOUTS[choral.choralId]){
+      TIMEOUTS[choral.choralId].seen = true
+      return
+    }
+
+    // make a new object for the chorals
+    TIMEOUTS[choral.choralId] = { seen:true }
+
+    console.log("new choral added " + choral.choralId)
+
+    // get the current value of the choral
+    client.hgetall(choral.choralId, (err, current_val) => {
+      if(err) return console.log(err.message)
+
+      var run_in = 0 
+      // if current_val == null then this is a brand new choral
+      // if current_val != null then there is alread some data for this choral
+      // in choral storm
+      if(current_val != null){
+        run_in = (choral.sampleRate * 1000) - ((new Date()).getTime() - current_val.device_timestamp)
+      }
+      run_in = Math.max(run_in, 0)
+
+      // this is admitidly a bit confusing, here is my thought
+      // this setTimeout is used to run the next iteration of the choral
+      // once that iteration is run, then the interval kicks in
+      // a timeline would look like this for a interval of 6 (underscores are seconds)
+      //
+      //      crash
+      // .. |_ _|_ _ |_ _|_ _ _ _ _ _|_ _ _ _ _ _|_ _ _ ...
+      //   last     now next      interval1   interval2
+      // so because we only have 2 seconds until the next time that
+      // the function should have ran we should run the function in 2 seconds then
+      // start the interval again on a regular interval
+      // if you have missed a interval then just run it now and restart the interval
+      //
+      //      crash
+      // .. |_ _|_ _ _ _|_ _ _|_ _ _ _ _ _|_ _ _ _ _ _|_ _ _  ...
+      //   last      missed now/next   interval1   interval2
+      //
+      // also note that you have to set the timeout here so that the
+      // the choral does not execute in the time between now and when the
+      // first timeout is fired
+      TIMEOUTS[choral.choralId].timeout_id = setTimeout(() => {
+        getChildrenAndRunChoral(choral)
+        TIMEOUTS[choral.choralId].timeout_id = setInterval(() => {
+          getChildrenAndRunChoral(choral)
+        }, choral.sampleRate * 1000)
+      }, run_in)
```

Comment:
I think I understand this, but why do we need to do it? Couldn't we just start the interval from the beginning?